internal/transport/http2_client.go

```diff
@@ -1023,6 +1023,14 @@ func (t *http2Client) Close(err error) {
 	case <-timer.C:
 		t.logger.Infof("Failed to write a GOAWAY frame as part of connection close after %s. Giving up and closing the transport.", goAwayLoopyWriterTimeout)
 	}
+	// Wait for the reader channel to close
+	select {
+	case <-t.readerDone:
```

Comment:
I think this needs to happen after the conn is closed, otherwise the reader will never exit.